packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,185 @@
+        return widget.kind === 'task';
+    }
+}
+
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig &&
+            (options.taskConfig.presentation === undefined || options.taskConfig.presentation.panel === undefined || options.taskConfig.presentation.panel === PanelKind.Shared);
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    @inject(TaskServer)
+    protected readonly taskServer: TaskServer;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            const finishedTaskId = event.taskId;
+            // find the terminal where the task ran, and mark it as "idle"
+            for (const terminal of this.getTaskTerminalWidgets()) {
+                if (terminal.taskId === finishedTaskId) {
+                    this.notifyTaskFinished(terminal);
+                    break;
+                }
+            }
+        });
+
+        this.terminalService.onDidCreateTerminal(async (widget: TerminalWidget) => {
+            const terminal = TaskTerminalWidget.is(widget) && widget;
+            if (terminal) {
+                const didConnectListener = terminal.onDidOpen(async () => {
+                    const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
+                    const tasksInfo = await this.taskServer.getTasks(context);
+                    const taskInfo = tasksInfo.find(info => info.terminalId === widget.terminalId);
+                    if (taskInfo) {
+                        const taskConfig = taskInfo.config;
+                        terminal.dedicated = !!taskConfig.presentation && !!taskConfig.presentation.panel && taskConfig.presentation.panel === PanelKind.Dedicated;
+                        terminal.taskId = taskInfo.taskId;
+                        terminal.taskConfig = taskConfig;
+                        terminal.busy = true;
+                    }
```

Comment:
are you sure that `else` is not required? this code is async - could it be that a task was finished in the meantime?

Reply:
you are right. i added the else block back in